File: JAVA/Soundex/Solution/src/main/java/com/kulturman/soundex/Soundex.java
package com.kulturman.soundex;

import java.util.HashMap;
import java.util.Map;

public class Soundex {
    private static final int ENCODED_WORD_MAX_SIZE = 4;

    public String encode(String word) {
        var encodedWord = encodeWordTail(word, wordHead(word));

        return addZerosPadding(encodedWord);
    }

    private String encodeWordTail(String word, String encodedWord) {
        StringBuilder result = new StringBuilder(encodedWord);

        var wordTail = wordTail(word);
        var lastEncodedLetter = "";

        for (int i = 0; i < wordTail.length(); i++) {
            if (isComplete(result))
                break;

            if (doesNotEncodeToSameValue(wordTail.charAt(i), lastEncodedLetter)) {
                result.append(encodeLetter(wordTail.charAt(i)));
            }

            lastEncodedLetter = encodeLetter(wordTail.charAt(i));
        }

        return result.toString();
    }

    private boolean doesNotEncodeToSameValue(char letter, String lastEncodedLetter) {
        return !lastEncodedLetter.equals(encodeLetter(letter));
    }

    private boolean isComplete(StringBuilder result) {
        return result.length() == ENCODED_WORD_MAX_SIZE;
    }

    private static String wordTail(String word) {
        return word.substring(1);
    }

    private static String wordHead(String word) {
        return String.valueOf(word.charAt(0)).toUpperCase();
    }

    String encodeLetter(char letter) {
        Map<String, String> soundexMap = new HashMap<>();

        soundexMap.put("b", "1");
        soundexMap.put("f", "1");
        soundexMap.put("p", "1");
        soundexMap.put("v", "1");

        soundexMap.put("c", "2");
        soundexMap.put("g", "2");
        soundexMap.put("j", "2");
        soundexMap.put("k", "2");
        soundexMap.put("q", "2");
        soundexMap.put("s", "2");
        soundexMap.put("x", "2");
        soundexMap.put("z", "2");

        soundexMap.put("d", "3");
        soundexMap.put("t", "3");

        soundexMap.put("l", "4");

        soundexMap.put("m", "5");
        soundexMap.put("n", "5");

        soundexMap.put("r", "6");

        return soundexMap.getOrDefault(String.valueOf(Character.toLowerCase(letter)), "");
    }

    String addZerosPadding(String encodedWord) {
        var zerosToAdd = ENCODED_WORD_MAX_SIZE - encodedWord.length();
        return encodedWord + "0".repeat(zerosToAdd);
    }
}

File: Typescript/Soundex/Solution/src/soundex.ts
export class Soundex {
    private readonly MAX_ENCODED_VALUE_LENGTH = 4;

    generate(word: string) {
        let encodedWord = word[0].toUpperCase();

        encodedWord += this.encodeWordTail(word.substring(1));

        return this.addZerosPad(encodedWord);
    }

    private encodeWordTail(word: string): string {
        let encodedValue = "";
        let lastEncodedValue: string | undefined = "";

        for (let letter of word) {
            if (this.encodingIsComplete(encodedValue))
                break;

            const currentEncodedValue = this.encodeLetter(letter);

            if (this.encodeLetter(letter) !== undefined && lastEncodedValue !== currentEncodedValue) {
                encodedValue += this.encodeLetter(letter);
            }

            lastEncodedValue = this.encodeLetter(letter);
        }
        return encodedValue;
    }

    private encodingIsComplete(encodedValue: string) {
        return encodedValue.length === this.MAX_ENCODED_VALUE_LENGTH - 1;
    }

    private addZerosPad(encodedWord: string) {
        const zerosToAdd = this.MAX_ENCODED_VALUE_LENGTH - encodedWord.length;

        return encodedWord + "0".repeat(zerosToAdd);
    }

    private encodeLetter(letter: string) {
        const encodingMap: Map<string, string> = new Map<string, string>([
            ['b', '1'], ['f', '1'], ['p', '1'], ['v', '1'],
            ['c', '2'], ['g', '2'], ['j', '2'], ['k', '2'], ['q', '2'],
            ['s', '2'], ['x', '2'], ['z', '2'],
            ['d', '3'], ['t', '3'],
            ['l', '4'],
            ['m', '5'], ['n', '5'],
            ['r', '6']
        ]);

        return encodingMap.get(letter.toLowerCase());
    }
}
